### PR TITLE
[Data][Docs] Convert "Custom datasource" guide to an example

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

The "Custom datasource" guide isn't really a guide -- it's an example of how to implement a specific custom datasource. Accordingly, I've moved the document to the examples section.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/examples/custom-datasource.rst [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
   This MongoDatasource guide below is for education only. For production use of MongoDB
   in Ray Data, see :ref:`Creating Dataset from MongoDB <dataset_mongo_db>`.
 
 Ray Data supports multiple ways to :ref:`create a dataset <loading_data>`,
-allowing you to easily ingest data of common formats from popular sources. However, if the 
-datasource you want to read from is not in the built-in list, don't worry, you can implement 
-a custom one for your use case. In this guide, we will walk you through how to build 
+allowing you to easily ingest data of common formats from popular sources. However, if the
+datasource you want to read from is not in the built-in list, don't worry, you can implement
+a custom one for your use case. In this guide, we will walk you through how to build
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:15]:
```suggestion
a custom one for your use case. This guide walks through building
```

File: doc/source/data/examples/custom-datasource.rst [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
 your own custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:16]:
```suggestion
a custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.
```

File: doc/source/data/examples/custom-datasource.rst [hunk 3]
```diff
@@ -40,8 +40,8 @@ By the end of the guide, you will have a ``MongoDatasource`` that you can use to
     which expresses document processing in a series of stages (e.g. match documents with a predicate, sort results, and then select a few fields).
     The execution results of the pipelines are used to create dataset.
 
-A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the 
-example here, let's call it ``MongoDatasource``. At a high level, it will have two 
+A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:43]:
```suggestion
A custom datasource is an implementation of :class:`~ray.data.Datasource`. In this
```

File: doc/source/data/examples/custom-datasource.rst [hunk 4]
```diff
@@ -40,8 +40,8 @@ By the end of the guide, you will have a ``MongoDatasource`` that you can use to
...
+example here, let's call it ``MongoDatasource``. At a high level, it will have two
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:44]:
```suggestion
example, it's called ``MongoDatasource``. At a high level, it has two
```

File: doc/source/data/examples/custom-datasource.rst [hunk 5]
```diff
@@ -103,10 +103,10 @@ to achieve this.
 
 .. literalinclude:: ./doc_code/custom_datasource.py
     :language: python
-    :start-after: __read_single_partition_start__ 
+    :start-after: __read_single_partition_start__
     :end-before: __read_single_partition_end__
 
-Once we have this building block, we can just apply it for each of the provided MongoDB 
+Once we have this building block, we can just apply it for each of the provided MongoDB
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:109]:
```suggestion
Once we have this building block, we apply it for each of the provided MongoDB
```

File: doc/source/data/examples/custom-datasource.rst [hunk 6]
```diff
@@ -159,17 +159,17 @@ and returns :ref:`their futures (object refs) <objects-in-ray>`.
...
 a ``MongoDatasource``.
 
 .. literalinclude:: ./doc_code/custom_datasource.py
     :language: python
     :start-after: __mongo_datasource_start__
     :end-before: __mongo_datasource_end__
 
-Now you can create a Dataset from and write back to MongoDB, just like 
+Now you can create a Dataset from and write back to MongoDB, just like
 any other datasource!
```
Comment (angelinalg) [inline doc/source/data/examples/custom-datasource.rst:173]:
```suggestion
any other datasource.
```

**Review by angelinalg — APPROVED** [review]

**Review by amogkam — COMMENTED** [review]
can we link to this example from all the places that discuss custom datasources? 

So creating datasets user guide, `Datasource` API ref, etc.

File: doc/source/data/examples/custom-datasource.rst [hunk 7]
```diff
@@ -1,19 +1,19 @@
...
   This MongoDatasource guide below is for education only. For production use of MongoDB
   in Ray Data, see :ref:`Creating Dataset from MongoDB <dataset_mongo_db>`.
 
 Ray Data supports multiple ways to :ref:`create a dataset <loading_data>`,
-allowing you to easily ingest data of common formats from popular sources. However, if the 
-datasource you want to read from is not in the built-in list, don't worry, you can implement 
-a custom one for your use case. In this guide, we will walk you through how to build 
-your own custom datasource, using `MongoDB <https://www.mongodb.com/docs/manual/introduction/>`__ as an example.
+allowing you to easily ingest data of common formats from popular sources. However, if the
+datasource you want to read from is not in the built-in list, don't worry, you can implement
```
Comment (amogkam) [inline doc/source/data/examples/custom-datasource.rst:14]: link to the built-in list?

**Review by amogkam — APPROVED** [review]

**Pull request merged**